### ISSUE TITLE
Dist versions do not work with Browserify

### ISSUE THREAD

**Issue by m90** [plotly/plotly.js#2902]
This is picking up: https://github.com/plotly/react-plotly.js/issues/93

I'm running into issues using the `dist` versions included in `plotly.js` and `plotly.js-dist` in a Browserify bundle.

I am basically just doing what the README tells me:

```js
var Plotly = require('plotly.js-dist');

console.log(Plotly);
```

When I try to Browserify this script it tells me:

```
Error: Cannot find module '../src/transforms/aggregate' from '/home/frederik/projects/plotly-browserify-issue/node_modules/plotly.js-dist'
```

When I try to load the dist bundle from the `plotly.js` package as seen for example [in `react-plotly.js`]( https://github.com/plotly/react-plotly.js/blob/6fe45a76d1542419fdbb9379941a631706bcf0a8/src/react-plotly.js#L2) I will see the following error:

```
Error: Cannot find module './core' from '/home/frederik/projects/plotly-browserify-issue/node_modules/plotly.js/dist'
```

I can however bundle plotly if I use `require('plotly.js')` as the main entry point.

I created an example repository for reproducing the issue over here: https://github.com/m90/plotly-browserify-issue

**Comment by m90** [comment]
It looks like this is currently expected and would need https://github.com/browserify/detective/pull/79 to land in order to work.

Other related issues:
- https://github.com/browserify/browserify/issues/1773
- https://github.com/browserify/browserify/issues/1847

which probably means that `dist` is not what Browserify users (or libraries that depend on plotly) should be using.

**Comment by etpinard** [comment]
Thanks very much for the detailed report. I was unaware that browserify couldn't bundle bundles it already bundled. 

It looks like https://github.com/browserify/detective/pull/79 is planned for browserify v17, so that's probably the best solution going forward.

At present though, if someone could give [`derequire`](https://www.npmjs.com/package/derequire) a shot, we will gladly merge that PR, but no plotly.js team member will dedicate time on this issue. That's mainly because bundling the "main" plotly.js package with browserify should [_just work_](https://github.com/plotly/plotly.js/blob/master/BUILDING.md#browserify) - that what we use internally afterall. In other words, for browserify users, `plotly.js-dist` only helps them save a few bytes of downloaded npm packages.

**Comment by m90** [comment]
> That's mainly because bundling the "main" plotly.js package with browserify should just work - that what we use internally afterall.

While this is 100% correct it still creates the issue of libraries that depend upon plotly to refer to the dist versions, thus making it impossible to use them with Browserify, see: https://github.com/plotly/react-plotly.js/issues/93 so in this case it'd not only be about the bytes.

**Comment by etpinard** [comment]
> see: plotly/react-plotly.js#93 so in this case it'd not only be about the bytes.

Oh right. I was unaware that `react-plotly.js` was already using `plotly.-js-dist`. Thanks for pointing this out!

**Comment by alexcjohnson** [comment]
Can we close this now that #2905 is merged?

**Comment by etpinard** [comment]
Yes. The fix in #2905 is compatible with browserify@17 which is upcoming.
